src/views/MainScreen.tsx

```diff
@@ -27,39 +31,77 @@ const styles = StyleSheet.create({
 });
 
-export default function MainScreen() {
-  const userId = 1;
-  const [exerciseLabel, exerciseLabelUpdate] = useState('My exercise');
-  const [exerciseCalories, exerciseCaloriesUpdate] = useState('100');
-
+const getTomorrowString = (): string => {
   const tomorrow = new Date();
   tomorrow.setDate(tomorrow.getDate() + 1);
-  const { data: exercises, refetch: refetchExercises } = useGetExerciseRecords({ userId, queryParams: { endDateTime: tomorrow.toISOString(), startDateTime: '0000-00-00 00:00:00' } });
-  const { mutate: postExercise } = usePostExerciseRecord({ userId });
-
-  const submitExercise = async () => {
-    const timestamp = new Date();
-
-    const exercise: ExerciseRecord = {
-      calories: parseInt(exerciseCalories, 10),
-      label: exerciseLabel,
-      timestamp: timestamp.toISOString()
-    };
-
-    await postExercise(exercise);
-    exerciseLabelUpdate('My exercise');
-    exerciseCaloriesUpdate('100');
-    refetchExercises();
-  };
+  tomorrow.setUTCHours(0);
+  tomorrow.setUTCMinutes(0);
+  tomorrow.setUTCSeconds(0);
+  tomorrow.setUTCMilliseconds(0);
+
+  // Get the ISO string and massage it into the format expected by the server (remove T separator between date and
+  // time, and remove fractional milliseconds and the timezone indicator)
+  return tomorrow.toISOString().replace('T', ' ').slice(0, -5);
+};
+
 
+export default function MainScreen() {
+  const [exerciseLabel, setExerciseLabel] = useState('');
+  const [exerciseCalories, setExerciseCalories] = useState('0');
+  const startTime = '0000-00-00 00:00:00';
+
+  // Get tomorrow's date. Only recompute if today's date changes
+  const today = new Date().getUTCDate();
+  const endTime = useMemo(() => getTomorrowString(), [today]);
+
+  const { data: exerciseList } = useQuery<GetExercisesQuery, GetExercisesQueryVariables>(GET_EXERCISE_DATA, {
+    variables: {
+      endTime, startTime
+    }
+  });
+
+  const [addExercise] = useMutation<AddExerciseMutation, AddExerciseMutationVariables>(ADD_EXERCISE_DATA, {
+    variables: {
+      input: {
+        userId: '1',
+        label: exerciseLabel,
+        calories: parseInt(exerciseCalories, 10)
+      }
+    },
+    update(cache, { data: newExerciseRecord }) {
+      // Read all exercises from local cache
+      const { user } = cache.readQuery<GetExercisesQuery, GetExercisesQueryVariables>({
+        query: GET_EXERCISE_DATA,
+        variables: { endTime, startTime }
+      });
+
+      // Add new exercise to this locally cached list and write it back to the cache
+      user.exerciseRecords.push(newExerciseRecord.createExerciseRecord);
+      cache.writeQuery<GetExercisesQuery, GetExercisesQueryVariables>({
+        query: GET_EXERCISE_DATA,
+        variables: { endTime, startTime },
+        data: { user }
+      });
+    }
+  });
+
+  const submitExercise = (): void => {
+    const caloriesInt = parseInt(exerciseCalories, 10);
+    if (exerciseLabel !== '' && caloriesInt > 0) {
+      addExercise();
```

Comment:
Kind of unfortunate that it's completely unclear what this is taking as input here. I can imagine this resulting in a lot of subtle bugs where values are updated in the wrong order. Is it at all possible to pass input as a parameter to `addExercise`? Otherwise, might it make sense to create a local `addExerciseInput` variable and set that before calling `addExercise` instead of relying on it picking up values from state during execution?

I'm concerned because it's not clear to me that addExercise would continue evaluating before state is updated in the following line, and I wonder whether there are any other weird race conditions that could crop up.

Reply:
You're right, and I don't know enough about the execution lifecycle of useMutate to know whether the vars get bound synchronously or not. Fortunately, it's also possible to just pass the variables object to the addExercise function instead of defining them on the mutation, which feels a bit better to me.